src/slotmap.jl

```diff
@@ -0,0 +1,230 @@
+    if startlen ≥ val_mask(isv)
+        error("SlotMapStructVector can store at most $(val_mask(isv)) elements.")
+    end
+    if iszero(slots_len)
+        push!(ID, startlen + 1)
+        setfield!(isv, :last_id, startlen + 1)
+    else
+        if iszero(free_head)
+            # push a slot to the end of slots
+            old_slots_capacity = length(slots)
+            if old_slots_capacity ≤ slots_len
+                if old_slots_capacity ≥ val_mask(isv)
+                    error("SlotMapStructVector is out of capacity")
+                end
+                # reallocate the slots
+                # avoid having a capacity larger than val_mask(isv)
+                new_slots_capacity = clamp(
+                    overallocation(old_slots_capacity),
+                    old_slots_capacity+1,
+                    val_mask(isv)
+                )
+                new_slots = typeof(slots)(undef, Int(new_slots_capacity))
+                copyto!(new_slots, slots)
+                setfield!(isv, :slots, new_slots)
+                slots = new_slots
+            end
+            # Start with generation 0
+            setfield!(isv, :slots_len, slots_len + 1)
+            new_id = Int64(slots_len) + 1
+            push!(ID, new_id)
+            setfield!(isv, :last_id, new_id)
+            slots[new_id] = (startlen + 1)%UInt64
+        else
+            # Pick a slot off the free list
+            @inbounds free_slot = slots[free_head]
+            next_free_head = (free_slot & val_mask(isv))%Int
+            old_gen = free_slot & gen_mask(isv)
+            next_gen = old_gen + (val_mask(isv) + 1)
+            new_slot = next_gen | UInt64(startlen + 1)
+            new_id = (next_gen | free_head%UInt64)%Int64
+            setfield!(isv, :free_head, next_free_head)
+            @inbounds slots[free_head] = new_slot
+            push!(ID, new_id)
+            setfield!(isv, :last_id, new_id)
+        end
+    end
+    unrolled_map(push!, values(comps)[2:end], t)
+    return isv
+end
+
+function Base.show(io::IO, ::MIME"text/plain", x::SlotMapStructVector{N, C}) where {N, C}
+    comps = getfield(x, :components)
+    sC = string(C)[13:end]
+    print("SlotMapStructVector{$sC")
+    return display(comps)
+end
+
+function Base.keys(isv::SlotMapStructVector)
+    return Keys(getfield(getfield(isv, :components), :ID))
+end
+
+@inline function Base.getindex(isv::SlotMapStructVector, id::Int64)
+    return IndexedView(id, id_guess_to_index(isv, id, id), isv)
+end
+
+function Base.in(a::IndexedView, isv::SlotMapStructVector)
+    getfield(a, :id) ∈ isv
+end
+function Base.in(id::Int64, isv::SlotMapStructVector)::Bool
+    comps = getfield(isv, :components)
+    slots_len, ID = getfield(isv, :slots_len), getfield(comps, :ID)
+    iszero(slots_len) && return id ∈ eachindex(ID)
+    if signbit(id)
+        return false
+    end
+    slots = getfield(isv, :slots)
+    slots_len = getfield(isv, :slots_len)
+    slot_idx = id & val_mask(isv)
```

Comment:
could you create a `slotidx(id)` function? e.g. here I think you forgot a `%Int` for perf

Reply:
sorry I should have said `slotidx(id, isv)`

Reply:
Also at https://github.com/Tortar/IndexedStructVectors.jl/pull/6/files#diff-574f88c83b20c2c2dbc2658985e196fc2cca368aefbe38e5945811e1e9be8559R62 I think you want to use that